### DISCUSSION TITLE
How to iterate self.db and self.ndb

### DISCUSSION THREAD

**Discussion by tszxxx** [evennia/evennia#2604] [Evennia Questions] [answered]
According to the documentation, I thought the way to do that should be: (I just modify the `Characters.py` file)
```
for a in self.db.all():
     ...
```
But in my case, `self.db.all` would be a list of attributes, so I need to use:
```
for attribute in self.db.all:
     attribute.key, attribute.value ...
```

But the weird thing happened for self.ndb.
`self.ndb.all` will return a list of keys(which is string type), not (key, value) pairs, and I cannot even use this way to access values:
```
        for key in self.ndb.all:
            print(self.ndb.__getattribute__(key))
```
it would throw:
```
21-12-17 20:14:42 [::]     print(self.ndb.__getattribute__(key))
21-12-17 20:14:42 [::] TypeError: 'NoneType' object is not callable
```
BTW, when I call `dir(self.ndb)`, it shows nothing
The only way now is to use `self.ndb.speed`, so I'm wondering if there is an easy way to iterate all (key, value) pairs for self.ndb

**Comment by tszxxx** [comment, marked as answer]
can use `getattr(self.ndb, 'speed')` instead

**Comment by Griatch** [comment]
The ndb-attributes are not properties on the .ndb handler in the normal sense, they echo how `Attributes` work, storing the value in a property `.value`. This is how you loop over them:
```python 
for nattr in self.ndb.all: 
    print(nattr.value)
```

The `.ndb` (and `.db`) are just convenience wrappers for the `.nattributes` (and `.attributes`) main handlers. So you can also do

```python
for nattr in self.nattributes.all():
    print(nattr.value)
```

**Reply by tszxxx** [reply]
Weird, when I use this in my code:
```
        for nattr in self.ndb.all:
            print(nattr.value)
```
I would get:
```
 File "/Users/zhenghangjie/PycharmProjects/muddev/mygame/typeclasses/properties.py", line 58, in __ndb_to_dict__
21-12-18 20:16:04 [::]     print(nattr.value)
21-12-18 20:16:04 [::] AttributeError: 'str' object has no attribute 'value'
```

**Comment by Griatch** [comment]
Maybe you overrode `ndb.all` with your own list at some point? It will allow you to do that, and will get back to the original behavior after you delete your custom `all` value.

**Reply by tszxxx** [reply]
Might be, but when I just do 
```
print(self.ndb.all)
```
it would return a list of keys like `["key1", "key2", "key3", ...]`, so I guess there is something wrong with my settings...
